packages/next/src/client/components/app-router.tsx

```diff
@@ -146,6 +147,76 @@ const createEmptyCacheNode = () => ({
   parallelRoutes: new Map(),
 })
 
+function useServerActionDispatcher(
+  changeByServerResponse: RouterChangeByServerResponse,
+  dispatch: React.Dispatch<ReducerActions>,
+  navigate: RouterNavigate
+) {
+  const serverActionDispatcher: ServerActionDispatcher = useCallback(
+    (actionPayload) => {
+      startTransition(() => {
+        dispatch({
+          ...actionPayload,
+          type: ACTION_SERVER_ACTION,
+          mutable: {},
+          navigate,
+          changeByServerResponse,
+        })
+      })
+    },
+    [changeByServerResponse, dispatch, navigate]
+  )
+  globalServerActionDispatcher = serverActionDispatcher
```

Comment:
Since this is exposed to the module's global and exports, previously it holds all states inside `Router` as it's defined inside. This fix affects both server and client.